### PR TITLE
JAMES-3534 API, Memory & contract test for identity storage

### PR DESCRIPTION
**Pull request by chibenwa**
-> Include a convenient API for updates, allowing updating several fields at once...
 
 -> Include a DAO layer dedicated to storage, and a repository layer ready to abstract default identities which will later be "patched".

### CODE REVIEW

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MDNSendMethod.scala [hunk 1]
```diff
@@ -76,7 +76,9 @@ class MDNSendMethod @Inject()(serializer: MDNSerializer,
                          invocation: InvocationWithContext,
                          mailboxSession: MailboxSession,
                          request: MDNSendRequest): SFlux[InvocationWithContext] =
-    identityResolver.resolveIdentityId(request.identityId, mailboxSession)
+    request.identityId.validate
+      .fold(e => SMono.error(new IllegalArgumentException("The IdentityId cannot be found", e)),
```
Comment (vttranlina) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MDNSendMethod.scala:81]: "The IdentityId is invalid"?

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/IdentityDAO.scala [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+  def list(user: Username): Publisher[Identity] = SFlux.merge(Seq(
+    customIdentityDao.list(user),
+    SMono.fromCallable(() => identityFactory.listIdentities(user))
+      .subscribeOn(Schedulers.elastic())
+      .flatMapMany(SFlux.fromIterable)))
+
+  def update(user: Username, identityId: IdentityId, identityUpdate: IdentityUpdate): Publisher[Unit] = customIdentityDao.update(user, identityId, identityUpdate)
+
+  def delete(username: Username, ids: Seq[IdentityId]): Publisher[Unit] = customIdentityDao.delete(username, ids)
```
Comment (vttranlina) [inline server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/IdentityDAO.scala]: Should we check Id is exists in `identityFactory`, if yes => throw an exception?

From https://jmap.io/spec-mail.html#identityset
```
Servers may wish to set this to false for the user’s username or other default address. Attempts to destroy an Identity with mayDelete: false will be rejected with a standard forbidden SetError.
```

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentityGetMethod.scala [hunk 3]
```diff
@@ -31,12 +32,12 @@ import org.apache.james.jmap.routes.SessionSupplier
...
 import org.apache.james.metrics.api.MetricFactory
 import play.api.libs.json.{JsError, JsObject, JsSuccess}
-import reactor.core.scala.publisher.SMono
+import reactor.core.scala.publisher.{SFlux, SMono}
 import reactor.core.scheduler.Schedulers
 
-class IdentityGetMethod @Inject() (identityFactory: IdentityFactory,
-                                  val metricFactory: MetricFactory,
-                                  val sessionSupplier: SessionSupplier) extends MethodRequiringAccountId[IdentityGetRequest] {
+class IdentityGetMethod @Inject() (identityFactory: IdentityRepository,
```
Comment (vttranlina) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentityGetMethod.scala:38]: `identityFactory` -> `identityRepository`, will more clear

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MDNSendMethod.scala (hunk 1 above)
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/MDNSendMethod.scala:81]: Here we match the ID format but not an UUID so we can confidently say the identity do not exist (which is a convention in JMAP)

Note that I wrote this not to break tests ;-)

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/IdentityDAO.scala (hunk 2 above)
Comment (chibenwa) [reply server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/identity/IdentityDAO.scala]: I let later tickets do this ;-)

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Identity.scala [hunk 4]
```diff
@@ -20,26 +20,22 @@
...
 
 case class IdentityName(name: String) extends AnyVal
 case class TextSignature(name: String) extends AnyVal
 case class HtmlSignature(name: String) extends AnyVal
 case class MayDeleteIdentity(value: Boolean) extends AnyVal
-case class IdentityId(id: Id)
-case class IdentityIds(ids: List[IdentityId]) {
-  def contains(identityId: IdentityId): Boolean = ids.contains(identityId)
-}
+case class IdentityId(id: UUID)
```
[the remark this answers is not in the document]
Comment (chibenwa) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Identity.scala]: UUID are easier to manage, easier to store than the generic format.

[the remark this answers is not in the document]
Comment (quantranhong1999) [reply server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/Identity.scala]: https://github.com/linagora/james-project/issues/4427#issuecomment-963823454

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MDNSendMethodContract.scala [hunk 5]
```diff
@@ -302,7 +303,7 @@ trait MDNSendMethodContract {
       `given`
         .body(request)
       .when
-        .post
+        .post.prettyPeek()
```
Comment (quantranhong1999) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MDNSendMethodContract.scala:306]: do not forget to delete :')

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MDNSendMethodContract.scala [hunk 6]
```diff
@@ -1035,7 +1036,7 @@ trait MDNSendMethodContract {
       `given`(buildBOBRequestSpecification(server))
         .body(request)
       .when
-        .post
+        .post.prettyPeek()
```
Comment (quantranhong1999) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MDNSendMethodContract.scala:1039]: idem

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MDNSendMethodContract.scala [hunk 7]
```diff
@@ -1214,7 +1215,7 @@ trait MDNSendMethodContract {
...
```
Comment (quantranhong1999) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MDNSendMethodContract.scala:1218]: idem

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MDNSendMethodContract.scala [hunk 8]
```diff
@@ -1374,7 +1375,7 @@ trait MDNSendMethodContract {
       `given`
         .body(request)
       .when
-        .post
+        .post.prettyPeek()
```
Comment (quantranhong1999) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MDNSendMethodContract.scala:1378]: idem

File: server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MDNSendMethodContract.scala [hunk 9]
```diff
@@ -1888,7 +1889,7 @@ trait MDNSendMethodContract {
     `given`
       .body(request)
     .when
-      .post
+      .post.prettyPeek()
```
Comment (quantranhong1999) [inline server/protocols/jmap-rfc-8621-integration-tests/jmap-rfc-8621-integration-tests-common/src/main/scala/org/apache/james/jmap/rfc8621/contract/MDNSendMethodContract.scala:1892]: idem

File: server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentityGetMethod.scala [hunk 10]
```diff
@@ -64,14 +65,17 @@ class IdentityGetMethod @Inject() (identityFactory: IdentityFactory,
...
-    SMono.fromCallable(() => identityFactory.listIdentities(mailboxSession))
+    SFlux(identityRepository.list(mailboxSession.getUser))
+      .collectSeq()
+      .map(_.toList)
       .map(request.computeResponse)
 }
 
-case class IdentityResolver @Inject()(identityFactory: IdentityFactory) {
-
+case class IdentityResolver @Inject()(identityFactory: IdentityRepository) {
```
Comment (quantranhong1999) [inline server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/method/IdentityGetMethod.scala:74]: identityFactory -> identityRepository ?

File: server/data/data-jmap/src/test/scala/org/apache/james/jmap/api/identity/CustomIdentityDAOContract.scala [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+  def testee(): CustomIdentityDAO
+
+  @Test
+  def listShouldReturnEmptyWhenNone(): Unit = {
+    assertThat(SFlux(testee().list(bob)).asJava().collectList().block())
+      .isEmpty()
+  }
+
+  @Test
+  def listShouldReturnSavedIdentities(): Unit = {
```
Comment (Arsnael) [inline server/data/data-jmap/src/test/scala/org/apache/james/jmap/api/identity/CustomIdentityDAOContract.scala:40]: Here we list one identity. Should we have a test to list multiple identities?

**Review by Arsnael — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/memory/identity/MemoryCustomIdentityDAO.scala [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+import scala.jdk.CollectionConverters._
+
+class MemoryCustomIdentityDAO extends CustomIdentityDAO {
+  private val table: Table[Username, IdentityId, Identity] = HashBasedTable.create
+
+  override def save(user: Username, creationRequest: IdentityCreationRequest): Publisher[Identity] =
+    SMono.fromCallable(() => IdentityId.generate)
+      .map(creationRequest.asIdentity)
+      .doOnNext(identity => table.put(user, identity.id, identity))
```
Comment (quantranhong1999) [inline server/data/data-jmap/src/main/scala/org/apache/james/jmap/memory/identity/MemoryCustomIdentityDAO.scala:37]: We should set default value for name, textSignature and htmlSignature in case they are empty (see https://jmap.io/spec-mail.html#identities)

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/model/Identity.scala [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+case class HtmlSignature(name: String) extends AnyVal
+case class MayDeleteIdentity(value: Boolean) extends AnyVal
+
+object IdentityId {
+  def generate: IdentityId = IdentityId(UUID.randomUUID())
+}
+case class IdentityId(id: UUID)
+
+case class Identity(id: IdentityId,
+                    name: IdentityName,
```
Comment (quantranhong1999) [inline server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/model/Identity.scala:37]: name: Option[IdentityName]

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/memory/identity/MemoryCustomIdentityDAO.scala (hunk 12 above)
Comment (quantranhong1999) [reply server/data/data-jmap/src/main/scala/org/apache/james/jmap/memory/identity/MemoryCustomIdentityDAO.scala:37]: Also, adding tests for default values when empty input?

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/model/Identity.scala (hunk 13 above)
Comment (quantranhong1999) [reply server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/model/Identity.scala:37]: Or we refine Identity like this
```scala
case class Identity(id: IdentityId,
                    name: IdentityName,
                    email: MailAddress,
                    replyTo: Option[List[EmailAddress]],
                    bcc: Option[List[EmailAddress]],
                    textSignature: TextSignature,
                    htmlSignature: HtmlSignature,
                    mayDelete: MayDeleteIdentity)
```
and set default value at IdentityCreationRequest::asIdentity ?

Comment (quantranhong1999) [reply server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/model/Identity.scala:37]: I see name, textSignature and htmlSignature always have value. So Optional here seems not necessary?

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/memory/identity/MemoryCustomIdentityDAO.scala (hunk 12 above)
Comment (chibenwa) [reply server/data/data-jmap/src/main/scala/org/apache/james/jmap/memory/identity/MemoryCustomIdentityDAO.scala:37]: Sure :+1:  missed this, good catch.

File: server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/model/Identity.scala (hunk 13 above)
Comment (chibenwa) [reply server/data/data-jmap/src/main/scala/org/apache/james/jmap/api/model/Identity.scala:37]: > name: Option[IdentityName]?

This cannot be nulll

**Review by Arsnael — APPROVED** [review]
